react/week3/todo-list-app/src/components/TodoList.js

```diff
@@ -0,0 +1,254 @@
+              name="deadline"
+              required
+            />
+            <br />
+          </label>
+          <button type="submit"> Add todo </button>
+        </form>
+      </div>
+    );
+  }
+}
+
+function FancyBorder(props) {
+  return <div className={"FancyBorder " + props.color}>{props.children}</div>;
+}
+
+class Todo extends React.Component {
+  state = {
+    isEditing: false,
+  };
+
+  toggleState = () => {
+    this.setState({
+      isEditing: !this.state.isEditing,
+    });
+  };
+
+  updateTodo = (event) => {
+    event.preventDefault();
+    this.props.editTodo(this.props.index, this.input);
+    this.toggleState();
+  };
+
+  renderEditMode = () => {
+    return (
+      <FancyBorder>
+        <li>
+          <form onSubmit={this.updateTodo}>
+            <input
+              type="text"
+              ref={(value) => {
+                this.input = value;
+              }}
+              defaultValue={this.props.todo.description}
+              required
+            />{" "}
+            | {this.props.todo.deadline}
+            <input
+              type="checkbox"
+              checked={this.props.todo.checked}
+              onChange={() => {
+                this.props.handleCheckbox(this.props.index);
+              }}
+            />
+            <button
+              className="delete-button"
+              onClick={() => {
+                this.props.deleteTodo(this.props.index);
+              }}
+            >
+              Delete
+            </button>
+            <button type="submit">Update</button>
+          </form>
+        </li>
+      </FancyBorder>
+    );
+  };
+
+  renderTodoList = () => {
+    return (
+      <FancyBorder>
+        <li
+          style={{
+            textDecoration: this.props.todo.checked ? "line-through" : "none",
+          }}
+        >
+          <label>
+            {this.props.todo.description} | {this.props.todo.deadline}
+          </label>
+          <input
+            type="checkbox"
+            checked={!!this.props.todo.checked}
+            onChange={() => {
+              this.props.handleCheckbox(this.props.index);
+            }}
+          />
+          <button
+            className="delete-button"
+            onClick={() => {
+              this.props.deleteTodo(this.props.index);
+            }}
+          >
+            Delete
+          </button>
+          <button
+            onClick={(event) => {
+              event.stopPropagation();
+              this.toggleState();
+            }}
+          >
+            Edit
+          </button>
+        </li>
+      </FancyBorder>
+    );
+  };
+  render() {
+    return (
+      <div>
+        {this.state.isEditing ? this.renderEditMode() : this.renderTodoList()}
+      </div>
+    );
+  }
+}
+
+function fetchTodo() {
+  const URL = `https://gist.githubusercontent.com/benna100/391eee7a119b50bd2c5960ab51622532/raw`;
+  return fetch(URL).then((response) => response.json());
```

Comment:
It would  practical to add `catch` to the `fetch` function to prevent the app from craching when there is a network error

Reply:
feedback implemented, thank you.